### PR TITLE
added case for rerun job for expried failed workflow

### LINKED ISSUE #31645: prow: handle the case of re-triggering an expired GitHub workflow
https://github.com/kubernetes/test-infra/pull/31132 added the functionality of re-triggering only failed jobs in a GitHub workflow. However, the [trigger](https://github.com/kubernetes/test-infra/blob/master/prow/plugins/trigger/generic-comment.go) plugin does not handle the cases when the failed workflow is expired (came up in discussion here https://github.com/kubernetes/test-infra/pull/30054#issuecomment-1637641500, cc @Priyankasaggu11929)

Workflows or jobs in workflows can only be re-run within `min(30 days, log retention period)` (or atleast that is my understanding based on https://docs.github.com/en/actions/managing-workflow-runs/re-running-workflows-and-jobs). Considering our repos have a 90 day log retention period, seems like we can only ever re-run jobs in a 30 day time-frame.

Regardless, if a repo uses prow to trigger GitHub workflows, `/retest` on expired workflows runs will yield no result from prow. We should probably make the bot post a comment indicating that one or more workflow run has expired. 

### Step 1: Modify the trigger logic here to include a comment if *at least* one of the runs is expired:

https://github.com/kubernetes/test-infra/blob/a3d4b4ec5b3814b5d4a9b4e65d770c0ec7f18bc0/prow/plugins/trigger/generic-comment.go#L160

The status code returned by an API call to re-run a failed, expired workflow run is `403` (unfortunately, GitHub does not document this), with an error message. There are 2 issues with trying to rely on this returned `403`:
1. `403` can indicate the need to retry the request. The Prow GitHub client will attempt a retry if status code of `403` is returned, but it will retry only if the error returned by the sending the request is `nil`

https://github.com/kubernetes/test-infra/blob/c758e68cdeb67314148a6be992957d092d63fcaa/prow/github/client.go#L946-L958

2. The error returned by executing a request to rerun an expired workflow run is non-nil, to verify I ran the following command:

```
❯ curl -L -X POST -s -w "%{http_code}\n" \                                                                                                                                     ─╯
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer MY_GITHUB_TOKEN" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/repos/kubernetes/utils/actions/runs/5577562429/rerun-failed-jobs
```

Note: you may not be able to run the exact same command above since you may not have write access on kubernetes/utils.

The output:
```
{
  "message": "Unable to retry this workflow run because it was created over a month ago",
  "documentation_url": "https://docs.github.com/rest/actions/workflow-runs#re-run-failed-jobs-from-a-workflow-run"
}
403
```

The same error is propagated into the Prow GitHub client as well.

As a result, the way the request is handled in the Prow client is that if an error is returned after all retries are done, it returns a status code of `0`, indicating that the client could not communicate with the server using HTTP (which is clearly not true). I've demonstrated this here with a unit test: https://github.com/MadhavJivrajani/test-infra/commit/38c5f900e22c844719e9786e0a0f64b56e705505 

The only way (for now) to know if we are trying to rerun an expired workflow run is to parse the error message itself. We can probably do:
```go
strings.Contains(err.Error(), "Unable to retry this workflow run because it was created over a month ago")
```

If the above condition is true, post a comment saying there are expired workflow runs.
It might also be worth including what can be done to mitigate this, one popular workaround I see online is to rebase or force push an empty commit.

### Step 2: Add tests

There aren't any tests for triggering GitHub workflows in general. If we do this, we need to ensure we have test coverage.

### Follow ups

Its worth looking into the returning of status code `0` considering this is a valid use case, but this is a follow up.
Also - note that we will end up retrying and backing off _every time_ we try to rerun an expired workflow run. Maybe there's a way to short-circuit that.

/sig contributor-experience testing

### PR DESCRIPTION
**Pull request by AvineshTripathi**
fixes #31645

### CODE REVIEW

File: prow/github/client.go
Comment (MadhavJivrajani) [inline prow/github/client.go:285]: Unnecessary diff

File: prow/plugins/trigger/generic-comment.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment.go:159]: Is there a reason we aren't running this in a separate goroutine anymore?

File: prow/plugins/trigger/generic-comment.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment.go:165]: We can't really return here because there might still be runs that can be successfully retried, if we return here, we basically don't do anything the moment we encounter an expired run.

File: prow/plugins/trigger/generic-comment.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment.go:163]: I'm assuming there's a link to be added in here?

Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment.go:163]: Let's also include the name of the expired workflow in the comment here, you can grab it from `run.Name`.

File: prow/plugins/trigger/generic-comment.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment.go:165]: Thinking more about the comment here, ideally we would want our comment to include the name of each expired workflow, something like this:
```
The following workflows could not be re-triggered because they are now expired (created over a month ago):
- name1
- name2
...
```

It is possible to comment as and when this happens and then keep editing the comment, but instead of doing that and to keep things simple, let's just keep track of the names of expired workflows in a slice and keep appending to the slice. 

Once we are out of the loop, check if the `len` of that slice is `>0` and if it is, then post a comment with all the names of the expired workflows.

**Review by MadhavJivrajani — COMMENTED** [review]
Thanks @AvineshTripathi! ❤️ 

Leaving a few initial comments.

File: prow/plugins/trigger/generic-comment.go
Comment (AvineshTripathi) [reply prow/plugins/trigger/generic-comment.go:159]: I was not sure if we want to Run the config.Presubmit in the case of expired workflow. Also I referred to this https://github.com/kubernetes/test-infra/blob/3aef270d485c0f747c9c5e310ef6e77d4ea5ba85/prow/plugins/trigger/generic-comment.go#L95 but not sure if that case applies to this too.

File: prow/plugins/trigger/generic-comment.go
Comment (AvineshTripathi) [reply prow/plugins/trigger/generic-comment.go:165]: can different workflows have different expired timings?

Comment (MadhavJivrajani) [reply prow/plugins/trigger/generic-comment.go:165]: @AvineshTripathi good question!
The only scenario that I had in mind while reviewing this was if an admin of the repo manually re triggered a particular workflow alone and not all. In that case, this is a scenario that can happen.

File: prow/plugins/trigger/generic-comment.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment.go:165]: @petr-muller / @stevekuznetsov do either of you remember why we are making calls to retrigger workflows in separate goroutines here?
Original PR: https://github.com/kubernetes/test-infra/pull/24726

Comment (petr-muller) [reply prow/plugins/trigger/generic-comment.go:165]: I don't, I wasn't that involved in that PR. Looking at the commits it seems like an opportunistic way to lower response latency because we only log errors but do not use them in the response? Goroutines were used right in the first [commit](https://github.com/kubernetes/test-infra/pull/24726/commits/1e4703ddd7ec3289eebd13800b3001d319c7adb9#diff-7271c7d0bdb752d0280ebeee09398796ec716e31f7174c98a4fbd6f496eba047R153-R157) so they were likley not added there as a result of a review (there are separate commits that addressed reviews).

Comment (petr-muller) [reply prow/plugins/trigger/generic-comment.go:165]: But yeah, that also means the `expiredWorklows` cannot be used this way, as we can easily hit L172 before the goroutines finish their client calls.

/hold

Comment (AvineshTripathi) [reply prow/plugins/trigger/generic-comment.go]: fixed that, i missed that thanks for spotting

Comment (AvineshTripathi) [reply prow/plugins/trigger/generic-comment.go]: do you think if I we go with this would we be able to ensure safety and also lower down the latency?

```
var expiredWorkflows []string
var wg sync.WaitGroup
var mu sync.Mutex

for _, run := range failedRuns {
    log := c.Logger.WithFields(logrus.Fields{
        "runID":   run.ID,
        "runName": run.Name,
        "org":     org,
        "repo":    repo,
    })
    wg.Add(1)
    go func(runName string, runID int64, log *logrus.Entry) {
        defer wg.Done()

        if err := c.GitHubClient.TriggerFailedGitHubWorkflow(org, repo, runID); err != nil {
            log.Errorf("attempt to trigger github run failed for %v: %v", runName, err)
            if strings.Contains(err.Error(), "Unable to retry this workflow run because it was created over a month ago") {
                mu.Lock()
                expiredWorkflows = append(expiredWorkflows, runName)
                mu.Unlock()
            }
        } else {
            log.Infof("successfully triggered action run")
        }
    }(run.Name, run.Name, log)
}

wg.Wait()

if len(expiredWorkflows) > 0 {
    comment := fmt.Sprintf(`The following workflows could not be re-triggered because they are now expired (created over a month ago):
- %s
`, strings.Join(expiredWorkflows, "\n- "))
    c.Logger.Infof("Commenting \"%s\".", comment)
    c.GitHubClient.CreateComment(org, repo, number, plugins.FormatResponseRaw(gc.Body, gc.HTMLURL, gc.User.Login, comment))
}

```

Comment (MadhavJivrajani) [reply prow/plugins/trigger/generic-comment.go]: @AvineshTripathi yes - that should be okay 👍🏼

Comment (MadhavJivrajani) [reply prow/plugins/trigger/generic-comment.go]: Please also take a look at the test failures.

**Comment by AvineshTripathi** [comment]
Hey @MadhavJivrajani please have a look at this PR. Thanks!

File: prow/plugins/trigger/generic-comment_test.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment_test.go:1511]: Thinking about this a little more -- I don't think you need to create a test server here. It would be easier if you modified the `FakeClient` that we have, specifically this method:
https://github.com/kubernetes/test-infra/blob/44aa1dbdbca77ac9e45d00c0c32681d547e6c289/prow/github/fakegithub/fakegithub.go#L1196-L1198

You can add a new field to the `FakeClient` called `TriggerFailedGitHubWorkflowError` similar to:
https://github.com/kubernetes/test-infra/blob/44aa1dbdbca77ac9e45d00c0c32681d547e6c289/prow/github/fakegithub/fakegithub.go#L142

And you can set this field each time you create a new FakeClient in each subtest, and based on if you're setting the error or not, you can return the desired error message in the fake implementation.

This way you also won't have to modify the `github.NewFakeClient()` method.

**Review by MadhavJivrajani — CHANGES_REQUESTED** [review]
Thanks for the ping @AvineshTripathi, this had dropped off my radar.
Just one comment, LGTM after that!

File: prow/plugins/trigger/generic-comment_test.go
Comment (AvineshTripathi) [reply prow/plugins/trigger/generic-comment_test.go:1511]: Hey I did write the test here is how it looks right now 
```
func TestRandom(t *testing.T) {
	var testCases = []struct {
		name            string
		triggerFailedGitHubWorkflowError error 
		runID           int
		expectedMessage string
	}{
		{
			name:            "Triggered workflow failed without err",
			runID:           1,
			triggerFailedGitHubWorkflowError: errors.New(``),
			expectedMessage: "",
		},
		{
			name:  "Triggered workflow failed due to expired workflow",
			triggerFailedGitHubWorkflowError: errors.New(`{"message": "Unable to retry this workflow run because it was created over a month ago","documentation_url": "https://docs.github.com/rest/actions/workflow-runs#re-run-failed-jobs-from-a-workflow-run"}`),
			runID: 2,
			expectedMessage: `{"message": "Unable to retry this workflow run because it was created over a month ago","documentation_url": "https://docs.github.com/rest/actions/workflow-runs#re-run-failed-jobs-from-a-workflow-run"}`,
		},
	}
	c := fakegithub.NewFakeClient()
	for _, testCase := range testCases {
		t.Run(testCase.name, func (t *testing.T) {
			c.TriggerFailedGitHubWorkflowError = testCase.triggerFailedGitHubWorkflowError
			err := c.TriggerFailedGitHubWorkflow("org", "repo", testCase.runID)
			if err != nil {
				if !strings.Contains(err.Error(), testCase.expectedMessage) {
					t.Errorf("Expected error message:\n%s\nGot:\n%s", testCase.expectedMessage, err.Error())
				}
			} else {
				if testCase.expectedMessage != "" {
					t.Errorf("Expected error message:\n%s\nGot: %s", testCase.expectedMessage, err)
				}
			} 
		})
	}
}
```

Comment (MadhavJivrajani) [reply prow/plugins/trigger/generic-comment_test.go:1511]: Go for it and update the PR! Its easier for me to leave comments once you do that. But your general approach looks good and is also what I had in mind 👍🏼

File: prow/github/client.go
Comment (MadhavJivrajani) [inline prow/github/client.go:1037]: Do we still need this now that we have this functionality in the fake client?

File: prow/github/fakegithub/fakegithub.go
Comment (MadhavJivrajani) [inline prow/github/fakegithub/fakegithub.go:1203]: We can probably replace this with just
```suggestion
	return f.TriggerFailedGitHubWorkflowError
```
A `nil` error is what we want to return in case of success anyway.

File: prow/plugins/trigger/generic-comment.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment.go:190]: Unnecessary diff.

File: prow/plugins/trigger/generic-comment_test.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment_test.go:1496]: We can remove this line entirely. "Without an error" implies that the error is `nil`, which is what we'd expect.

File: prow/plugins/trigger/generic-comment_test.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment_test.go:1490]: We don't need `runID` anymore.

File: prow/plugins/trigger/generic-comment_test.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment_test.go:1509]: Hmm
I think let's create the client newly each time and assign the error value? We prevent races this way (right now we probably don't have any other test making use of `TriggerFailedGitHubWorkflow` but in the future there might be, and that might run in parallel, so let's avoid that).

File: prow/plugins/trigger/generic-comment_test.go
Comment (MadhavJivrajani) [inline prow/plugins/trigger/generic-comment_test.go:1519]:
```go
if c.triggerFailedGitHubWorkflowError == nil && err != nil {
    t.Errorf("Unexpected error..."...)
}
if c.triggerFailedGitHubWorkflowError != nil  && err == nil{
    t.Errorf("expected error, did not get one..."...)
}
if c.triggerFailedGitHubWorkflowError != nil  && err.Error() != expectedMessage  {
    t.Errorf("expected error, did not get one..."...)
}

Comment (MadhavJivrajani) [reply prow/plugins/trigger/generic-comment_test.go:1519]: You can add in the required formatting and debugging info here ^

**Review by MadhavJivrajani — CHANGES_REQUESTED** [review]
Last set of comments, thanks for working on this! 

Please also squash commits after this.

*Note:* this is a followup, but I think we should also be testing the exact comments that are going to be posted as well (similar to the lgtm and approve plugins).

File: prow/github/client.go
Comment (AvineshTripathi) [reply prow/github/client.go:1037]: Didn't get you here, you mean we should allow retry to happen the usual way and let things return in normally cause we have handled the err at caller level i.e. where we call `TriggerFailedGitHubWorkflow`?

Comment (MadhavJivrajani) [reply prow/github/client.go:1037]: Oh nvm, I misunderstood this change. I agree with you, we shouldn't retry if its expired 👍🏼 
I would however prefer if this change got a second set of eyes, @petr-muller, could you maybe take a look here? 

If we are modifying how the retry happens, we should have an added test cases for the prow client that shows that for an error message that we are parsing for, we get a `403` and a non-nil error.

You can probably reuse most of this test to achieve that: https://github.com/MadhavJivrajani/test-infra/commit/38c5f900e22c844719e9786e0a0f64b56e705505

Comment (AvineshTripathi) [reply prow/github/client.go:1037]: waiting for final call on this and will update the PR

Comment (MadhavJivrajani) [reply prow/github/client.go:1037]: @AvineshTripathi pls feel free to update the PR, its always better to have the changes in front of you for ease of review.

**Comment by MadhavJivrajani** [comment]
@AvineshTripathi please fix the test failures and squash commits

**Comment by AvineshTripathi** [comment]
@MadhavJivrajani  not sure if I can do anything to fix this `Pod scheduling timeout`

**Comment by MadhavJivrajani** [comment]
/retest 
@AvineshTripathi don't worry about it! I think the test cluster should be fine now.

**Comment by jihoon-seo** [comment]
Hello @AvineshTripathi, the Prow source code has been moved to https://github.com/kubernetes-sigs/prow on April 9, 2024.
Please consider migrate this PR (by opening a new one in there).
Thanks!

/remove-area config jobs provider/azure provider/gcp release-eng
/remove-sig cloud-provider cluster-lifecycle release storage

**Pull request closed without merging**
